### DISCUSSION TITLE
Weekly Update Contract Runtime 2020-12-17

### DISCUSSION THREAD

**Discussion by ailisp** [near/nearcore#3752] [General]
Willem:
- Wasmer compile with stable rust, Piotr find some issue
- Profiling PR ready to merge, publish deps for profiling crate
- PR in sdk-rs about profiling data
- Another PR in near-sdk-as add singleton, some naming and parameter of the decorator need decision
- simulator has a memory issue in AVL tree kv pair found by Mike. Might be due to create trie, or compilation doing more than once, still investigating

Bo:
- decode and re-encode args to evm runner with eth-abi
- replace wasmer 0.17 aritfact serializer with borsh

Nikolay:
- finished model for io cost

Evgeny:
- Fixed P0 performance issue due to feature leak.

Some disucssions on AS, TS and Rust
- Max: AS not production ready, good to have but not on the level on rust
- Willem: not as mature, good for prototype, educational, should still invest on.
- Advertising AS as TS might be wrong, it easy for TS developer get started in, but not good for port all TS libraries
- Nikolay: we should do fix-only on as-sdk
- Bo: given AS we have existing as-sdk, doc and other developer infra, it's probably still in a condition that's better than other lang and worse than rust. If we still supporting one other language other than rust, it should be AS and put 20% bandwidth of SDK on it.
- May need runtime team and whole organization discussion

**Comment by mikedotexe** [comment]
I'll add a note here that @thisisjoshford came over to work a couple weeks ago and we spent hours having issues with the TreeMap implementation in AssemblyScript. Turns out, thanks to Josh and @willemneal they nailed down the issue and Josh opened this issue:
https://github.com/AssemblyScript/assemblyscript/issues/1582
This is a pretty serious, foundational problem in my opinion, and I'm glad we caught it instead of developers. It also means that TreeMap had this critical issue. This makes me wonder what other unknowns are there, lurking. I agree that this is "good for prototype"
In my opinion, as someone who is not as strong of a developer as a lot of the folks at NEAR, I think we could send a message that while Rust may look scary if you peer into the nearcore codebase, building smart contracts with Rust really isn't, thanks to near-sdk-rs. And for folks wanting to build smart contracts with AssemblyScript just give them a warnings that we sometimes find issues and it's a growing language, etc. I don't feel like we have that message in our documentation and examples.
Lastly, I feel bad for Willem as he's pretty much siloed on AssemblyScript. I think if we decide to invest it in, we might consider adding to the bus factor.

**Reply by MaxGraey** [reply]
I'm wondering what do you think about this recently unsoundness bug in Rust which existed since 2013:
https://blog.rust-lang.org/2020/07/16/Rust-1.45.0.html#fixing-unsoundness-in-casts

Another meaningful bugs which still opened:

https://github.com/rust-lang/rust/issues/28728
https://github.com/rust-lang/rust/issues/78022
https://github.com/rust-lang/rust/issues/60076
https://github.com/rust-lang/rust/issues/25860
https://github.com/rust-lang/rust/issues/80303

I just want to show that the presence of bugs, even the most serious and old, does not yet make a language or a compiler non-production ready

**Reply by MaxGraey** [reply]
Another story:
https://engineering.skroutz.gr/blog/uncovering-a-24-year-old-bug-in-the-linux-kernel
